mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -335,32 +644,102 @@ bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
 }
 
 //===----------------------------------------------------------------------===//
-// AlwaysInlinerPass
+// Inliner
 //===----------------------------------------------------------------------===//
 
-void AlwaysInlinerPass::runOnOperation() {
-  CallGraph &CG = getAnalysis<CallGraph>();
-  auto *Ctx = &getContext();
+bool Inliner::shouldInline(ResolvedCall &ResolvedCall,
+                           Optional<CGUseList> Uses) const {
+  if (isRecursiveCall(ResolvedCall))
+    return false;
 
-  // The inliner should only be run on operations that define a symbol table,
-  // as the call graph will need to resolve references.
-  Operation *Op = getOperation();
-  if (!Op->hasTrait<OpTrait::SymbolTable>()) {
-    Op->emitOpError() << " was scheduled to run under the inliner, but does "
-                         "not define a symbol table";
-    return signalPassFailure();
+  FunctionOpInterface Callee = getCalledFunction(ResolvedCall.Call);
+  NamedAttrList FnAttrs(Callee->getAttrDictionary());
+  Optional<NamedAttribute> PassThroughAttr = FnAttrs.getNamed("passthrough");
+
+  // Inline a function if it has an attribute suggesting that inlining is
+  // desirable.
+  if (PassThroughAttr) {
+    bool ShouldInline = llvm::any_of(
+        PassThroughAttr->getValue().cast<ArrayAttr>(), [](Attribute Attr) {
+          return Attr.isa<StringAttr>() &&
+                 ((Attr.cast<StringAttr>() == "alwaysinline") ||
+                  (Attr.cast<StringAttr>() == "inlinehint"));
+        });
+    if (ShouldInline)
+      return true;
   }
 
+  // Inline a function if inlining makes it dead.
+  if (Uses.has_value())
+    return Uses->hasOneUse(ResolvedCall.TgtNode);
+
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// InlinePass
+//===----------------------------------------------------------------------===//
+
+void InlinePass::runOnOperation() {
+  if (!checkForSymbolTable(*getOperation()))
+    return signalPassFailure();
+
+  MLIRContext *Ctx = &getContext();
+  CallGraph &CG = getAnalysis<CallGraph>();
   SymbolTableCollection SymTable;
-  AlwaysInliner Inliner(Ctx, CG, SymTable);
-  LogicalResult Res = runTransformOnSCCs(CG, [&](CallGraphSCC &SCC) {
-    return Inliner::inlineSCC(Inliner, SCC, Ctx);
-  });
+  CGUseList UseList(getOperation(), CG, SymTable);
+
+  std::unique_ptr<InlinerBase> Inliner = nullptr;
+  switch (InlineMode) {
+  case sycl::InlineMode::AlwaysInline:
+    Inliner = std::make_unique<class AlwaysInliner>(Ctx, CG, SymTable);
+    break;
+  case sycl::InlineMode::Simple:
+    Inliner = std::make_unique<class Inliner>(Ctx, CG, SymTable);
```

Comment:
optional:
```suggestion
    Inliner = std::make_unique<class SimpleInliner>(Ctx, CG, SymTable);
```

Reply:
Not for this PR, but I am starting to think that we could just have one Inliner class implementing the `shouldInline` member function based on heuristics. Or we could have an `Inliner` class that takes a `InlineHeuristic` object. That object would supply the heuristic to use.

Let's keep it like it is in this PR.